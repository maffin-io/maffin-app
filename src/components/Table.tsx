import React from 'react';
import {
  ColumnDef,
  TableOptions,
  useReactTable,
  flexRender,
  getCoreRowModel,
  getSortedRowModel,
  getExpandedRowModel,
  ColumnSort,
  ExpandedState,
  getPaginationRowModel,
} from '@tanstack/react-table';
import classNames from 'classnames';
import { FaSortDown, FaSortUp } from 'react-icons/fa';

import Pagination from '@/components/Pagination';

export type TableProps<T extends object> = {
  id: string,
  columns: ColumnDef<T>[],
  data: T[],
  initialSort?: ColumnSort,
  showHeader?: boolean,
  showPagination?: boolean,
  tdClassName?: string,
  isExpanded?: boolean,
} & Partial<TableOptions<T>>;

export default function Table<T extends object = {}>(
  {
    id,
    initialSort,
    showHeader = true,
    showPagination = false,
    tdClassName = 'px-6 py-4',
    isExpanded = false,
    ...props
  }: TableProps<T>,
): JSX.Element {
  const tableConfig: TableOptions<T> = {
    getCoreRowModel: getCoreRowModel(),
    getSortedRowModel: getSortedRowModel(),
    getExpandedRowModel: getExpandedRowModel(),
    enableExpanding: true,
    initialState: {
      ...props.initialState,
      sorting: (initialSort && [initialSort]) || undefined,
      expanded: isExpanded as ExpandedState,
    },
    ...props,
  };

  if (showPagination && !props.manualPagination) {
    tableConfig.getPaginationRowModel = getPaginationRowModel();
  }

  const table = useReactTable(tableConfig);

  return (
    <>
      <div className="relative overflow-hidden rounded-md">
        <table id={id} className="w-full text-sm text-left">
          {
            showHeader
            && (
              <thead className="bg-white dark:bg-dark-700">
                {table.getHeaderGroups().map((headerGroup) => (
                  <tr key={headerGroup.id}>
                    {headerGroup.headers.map(header => (
                      <th
                        key={header.id}
                        scope="col"
                        className="px-6 py-3"
                        onClick={header.column.getToggleSortingHandler()}
                      >
                        <div className="flex">
                          {flexRender(header.column.columnDef.header, header.getContext())}
                          {(
                            header.column.columnDef.header !== ''
                              && header.column.getCanSort()
                              && (
                                <div className="flex items-center cursor-pointer">
                                  <FaSortUp
                                    className={
                                      classNames(
                                        {
                                          'opacity-100': header.column.getIsSorted() === 'asc',
                                          'opacity-50': header.column.getIsSorted() !== 'asc',
                                        },
                                        'absolute',
                                      )
                                    }
                                  />
                                  <FaSortDown
                                    className={
                                      classNames(
                                        {
                                          'opacity-100': header.column.getIsSorted() === 'desc',
                                          'opacity-50': header.column.getIsSorted() !== 'desc',
                                        },
                                        'absolute',
                                      )
                                    }
                                  />
                                </div>
                              )
                          )}
                        </div>
                      </th>
                    ))}
                  </tr>
                ))}
              </thead>
            )
          }

          <tbody>
            {table.getRowModel().rows.map(row => (
              <tr key={row.id} className="border-b border-white/70 dark:border-dark-700">
                {row.getVisibleCells().map(cell => (
                  <td key={cell.id} className={tdClassName}>
                    {flexRender(cell.column.columnDef.cell, cell.getContext())}
                  </td>
                ))}
              </tr>
            ))}
          </tbody>
        </table>
      </div>

      {
        showPagination
        && <Pagination<T> table={table} />
      }
    </>
  );
}
