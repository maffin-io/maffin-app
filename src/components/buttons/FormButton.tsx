import React from 'react';

import { DataSourceContext } from '@/hooks';
import Modal, { ModalRef } from '@/components/ui/Modal';

export interface FormButtonProps extends Omit<
React.ButtonHTMLAttributes<HTMLButtonElement>,
'children'
> {
  buttonContent: string | React.ReactElement,
  modalTitle: string,
  children: React.ReactElement,
}

export default function FormButton({
  buttonContent,
  modalTitle,
  children,
  className = 'btn btn-primary',
  ...props
}: FormButtonProps): React.JSX.Element {
  const { save } = React.useContext(DataSourceContext);
  const modalRef = React.useRef<ModalRef>(null);

  return (
    <Modal
      ref={modalRef}
      className="modal card"
      triggerContent={buttonContent}
      triggerProps={{ ...props, className }}
      showClose
    >
      <span>{modalTitle}</span>
      <div>
        {
          React.cloneElement(
            children,
            {
              // @ts-ignore
              onSave: (e: unknown) => {
                // @ts-ignore
                children.props.onSave?.(e);
                save();
                modalRef.current?.closeModal();
              },
            },
          )
        }
      </div>
    </Modal>
  );
}
