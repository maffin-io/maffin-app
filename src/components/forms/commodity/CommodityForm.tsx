import React from 'react';
import { useForm, Controller } from 'react-hook-form';
import { classValidatorResolver } from '@hookform/resolvers/class-validator';
import { SingleValue } from 'react-select';

import { Commodity } from '@/book/entities';
import Selector from '@/components/selectors/Selector';

const resolver = classValidatorResolver(Commodity, { validator: { stopAtFirstError: true } });

export type FormValues = {
  namespace: string,
  mnemonic: string,
};

export type CommodityFormProps = {
  onSave: Function,
};

const BASE_CURRENCIES = [
  { label: 'EUR' },
  { label: 'USD' },
  { label: 'SGD' },
  { label: 'GBP' },
];

export default function CommodityForm({ onSave }: CommodityFormProps): JSX.Element {
  const form = useForm<FormValues>({
    mode: 'onChange',
    resolver,
  });

  return (
    <form onSubmit={form.handleSubmit((data) => onSubmit(data, onSave))}>
      <fieldset className="text-sm my-5">
        <Controller
          control={form.control}
          name="mnemonic"
          render={({ field, fieldState }) => (
            <>
              <Selector<{ label: string }>
                id="mnemonicInput"
                labelAttribute="label"
                options={BASE_CURRENCIES}
                onChange={(newValue: SingleValue<{ label: string }> | null) => {
                  field.onChange(newValue?.label);
                }}
                placeholder="Choose your currency"
                isClearable={false}
              />
              <p className="invalid-feedback">{fieldState.error?.message}</p>
            </>
          )}
        />
      </fieldset>

      <input
        {...form.register(
          'namespace',
        )}
        hidden
        value="CURRENCY"
        type="text"
      />

      <div className="flex w-full justify-center">
        <button className="btn btn-primary" type="submit" disabled={!form.formState.isValid}>
          Save
        </button>
      </div>
    </form>
  );
}

async function onSubmit(data: FormValues, onSave: Function) {
  const mainCommodity = await Commodity.create({ ...data }).save();
  await onSave(mainCommodity);
}
