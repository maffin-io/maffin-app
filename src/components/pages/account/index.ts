export { default as TotalLineChart } from './TotalLineChart';
export { default as SplitsHistogram } from './SplitsHistogram';
export { default as TransactionsTable } from './TransactionsTable';
export { default as Header } from './Header';
export { default as AccountInfo } from './AccountInfo';
export { default as InvestmentInfo } from './InvestmentInfo';
export { default as AssetInfo } from './AssetInfo';
