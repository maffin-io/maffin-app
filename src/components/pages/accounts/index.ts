export { default as AccountsTable } from './AccountsTable';
export { default as MonthlyTotalHistogram } from './MonthlyTotalHistogram';
export { default as NetWorthPie } from './NetWorthPie';
export { default as LatestTransactions } from './LatestTransactions';
