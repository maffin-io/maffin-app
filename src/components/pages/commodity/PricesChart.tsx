import React from 'react';

import Line from '@/components/charts/Line';
import { moneyToString } from '@/helpers/number';
import { Price } from '@/book/entities';

export type TotalLineChartProps = {
  prices: Price[],
};

export default function PricesChart({
  prices,
}: TotalLineChartProps): JSX.Element {
  const currency = (prices.length && prices[0].currency.mnemonic) || '';

  const data: { x: number, y: number }[] = [];
  prices.forEach(price => {
    data.push({
      x: price.date.toMillis(),
      y: price.value,
    });
  });

  return (
    <Line
      data={{
        datasets: [
          {
            data,
            pointStyle: 'circle',
          },
        ],
      }}
      options={{
        plugins: {
          datalabels: {
            display: false,
          },
          legend: {
            display: false,
          },
          tooltip: {
            backgroundColor: '#323b44',
            callbacks: {
              label: (ctx) => `${moneyToString(Number(ctx.parsed.y), currency)}`,
            },
          },
        },
        scales: {
          x: {
            type: 'time',
            time: {
              unit: 'day',
              round: 'day',
              tooltipFormat: 'dd MMMM yyyy',
            },
            grid: {
              display: false,
            },
          },
          y: {
            border: {
              display: false,
            },
            ticks: {
              maxTicksLimit: 10,
              callback: (value) => moneyToString(value as number, currency),
            },
          },
        },
      }}
    />
  );
}
