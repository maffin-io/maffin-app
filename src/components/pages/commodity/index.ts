export { default as PricesChart } from './PricesChart';
