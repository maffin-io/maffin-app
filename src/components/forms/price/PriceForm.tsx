import React from 'react';
import { DateTime } from 'luxon';
import { useForm, Controller } from 'react-hook-form';
import { classValidatorResolver } from '@hookform/resolvers/class-validator';
import { mutate } from 'swr';
import classNames from 'classnames';

import { Price } from '@/book/entities';
import { CommoditySelector } from '@/components/selectors';
import type { Commodity } from '@/book/entities';
import { FormValues } from './types';

const resolver = classValidatorResolver(Price, { validator: { stopAtFirstError: true } });

export type PriceFormProps = {
  action?: 'add' | 'update' | 'delete',
  onSave?: Function,
  defaultValues?: Partial<FormValues>,
  hideDefaults?: boolean,
};

export default function PriceForm({
  action = 'add',
  defaultValues,
  onSave = () => {},
  hideDefaults = false,
}: PriceFormProps): JSX.Element {
  const form = useForm<FormValues>({
    defaultValues,
    mode: 'onChange',
    resolver,
  });

  const { errors } = form.formState;
  const disabled = action === 'delete';

  return (
    <form onSubmit={form.handleSubmit((data) => onSubmit(data, action, onSave))}>
      <fieldset className="text-sm my-5">
        <label htmlFor="dateInput" className="inline-block mb-2">Date</label>
        <input
          id="dateInput"
          disabled={disabled}
          className="block w-full m-0"
          {...form.register('date')}
          type="date"
        />
        <p className="invalid-feedback">{errors.date?.message}</p>
      </fieldset>

      <fieldset
        className={classNames(
          'text-sm my-5',
          {
            hidden: hideDefaults && defaultValues?.fk_commodity,
          },
        )}
      >
        <label htmlFor="commodityInput" className="inline-block mb-2">From</label>
        <Controller
          control={form.control}
          name="fk_commodity"
          render={({ field, fieldState }) => (
            <>
              <CommoditySelector
                id="commodityInput"
                placeholder="Convert from"
                onChange={field.onChange}
                isDisabled={disabled || action !== 'add'}
                defaultValue={defaultValues?.fk_commodity as Commodity}
              />
              <p className="invalid-feedback">{fieldState.error?.message}</p>
            </>
          )}
        />
      </fieldset>

      <div className="grid grid-cols-12 text-sm my-5 gap-2">
        <fieldset
          className={classNames(
            'col-span-6',
            {
              hidden: hideDefaults && defaultValues?.fk_currency,
            },
          )}
        >
          <label htmlFor="currencyInput" className="inline-block mb-2">To</label>
          <Controller
            control={form.control}
            name="fk_currency"
            render={({ field, fieldState }) => (
              <>
                <CommoditySelector
                  id="currencyInput"
                  placeholder="Convert to"
                  onChange={field.onChange}
                  namespace="CURRENCY"
                  isDisabled={disabled || action !== 'add'}
                  defaultValue={defaultValues?.fk_currency as Commodity}
                />
                <p className="invalid-feedback">{fieldState.error?.message}</p>
              </>
            )}
          />
        </fieldset>

        <fieldset className="col-span-6">
          <label htmlFor="priceInput" className="inline-block text-sm mb-2">Price</label>
          <input
            id="priceInput"
            disabled={disabled}
            className="w-full m-0"
            {...form.register('value')}
            type="text"
          />
          <p className="invalid-feedback">{errors.value?.message}</p>
        </fieldset>
      </div>

      <div className="flex w-full justify-center mt-5">
        <button
          className={classNames(
            'btn capitalize',
            {
              'btn-primary': action === 'add',
              'btn-warn': action === 'update',
              'btn-danger': action === 'delete',
            },
          )}
          type="submit"
          disabled={Object.keys(errors).length > 0}
        >
          {action}
        </button>
      </div>
    </form>
  );
}

async function onSubmit(
  data: FormValues,
  action: 'add' | 'update' | 'delete',
  onSave: Function,
) {
  const price = Price.create({
    ...data,
    date: DateTime.fromISO(data.date),
    guid: data.guid || undefined,
  });

  if (action === 'add' || action === 'update') {
    await Price.upsert(
      [price],
      {
        conflictPaths: ['fk_commodity', 'fk_currency', 'date'],
      },
    );
  } else if (action === 'delete') {
    await Price.remove(price);
  }
  mutate(`/api/prices/${price.commodity.guid}`);

  onSave(price);
}
