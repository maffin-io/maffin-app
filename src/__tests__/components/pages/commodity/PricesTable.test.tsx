import React from 'react';
import { DateTime } from 'luxon';
import {
  render,
  screen,
} from '@testing-library/react';

import Table from '@/components/Table';
import FormButton from '@/components/buttons/FormButton';
import PriceForm from '@/components/forms/price/PriceForm';
import { PricesTable } from '@/components/pages/commodity';
import type { Price } from '@/book/entities';

jest.mock('@/components/Table', () => jest.fn(
  () => <div data-testid="Table" />,
));

jest.mock('@/components/buttons/FormButton', () => jest.fn(
  (props: React.PropsWithChildren) => (
    <div data-testid="FormButton">
      {props.children}
    </div>
  ),
));

jest.mock('@/components/forms/price/PriceForm', () => jest.fn(
  () => <div data-testid="PriceForm" />,
));

describe('PricesTable', () => {
  it('creates empty table with expected params', async () => {
    render(<PricesTable prices={[]} />);

    await screen.findByTestId('Table');
    expect((Table as jest.Mock)).toHaveBeenLastCalledWith(
      {
        id: 'prices-table',
        initialSort: {
          desc: true,
          id: 'date',
        },
        pageSize: 7,
        data: [],
        columns: [
          {
            header: 'Date',
            id: 'date',
            accessorFn: expect.any(Function),
            cell: expect.any(Function),
          },
          {
            header: 'Rate',
            enableSorting: false,
            accessorFn: expect.any(Function),
          },
          {
            header: 'Actions',
            enableSorting: false,
            cell: expect.any(Function),
          },
        ],
      },
      {},
    );
    expect(
      (Table as jest.Mock).mock.calls[0][0].columns[0].accessorFn({ date: DateTime.fromISO('2023-01-01') }),
    ).toEqual(1672531200000);
    expect(
      (Table as jest.Mock).mock.calls[0][0].columns[1].accessorFn({
        value: 100.50,
        fk_currency: {
          mnemonic: 'USD',
        },
      }),
    ).toEqual('$100.50');
  });

  it('forwards prices as data', async () => {
    const prices = [
      {
        guid: '1',
        value: 100,
      } as Price,
    ];

    render(<PricesTable prices={prices} />);

    expect((Table as jest.Mock)).toHaveBeenCalledWith(
      expect.objectContaining({
        data: prices,
      }),
      {},
    );
  });

  it('renders Actions column as expected', async () => {
    const prices = [
      {
        guid: '1',
        value: 100,
        date: DateTime.fromISO('2023-01-01'),
      } as Price,
    ];

    render(<PricesTable prices={prices} />);

    await screen.findByTestId('Table');
    const actionsCol = (Table as jest.Mock).mock.calls[0][0].columns[2];

    expect(actionsCol.cell).not.toBeUndefined();

    const { container } = render(
      // @ts-ignore
      actionsCol.cell({
        row: {
          original: prices[0],
        },
      }),
    );

    expect(FormButton).toHaveBeenNthCalledWith(
      1,
      expect.objectContaining({
        className: 'link',
        id: 'edit-price',
        modalTitle: 'Edit price',
      }),
      {},
    );
    expect(PriceForm).toHaveBeenNthCalledWith(
      1,
      {
        action: 'update',
        defaultValues: {
          date: '2023-01-01',
          fk_commodity: undefined,
          fk_currency: undefined,
          guid: '1',
          value: 100,
        },
      },
      {},
    );
    expect(FormButton).toHaveBeenNthCalledWith(
      2,
      expect.objectContaining({
        className: 'link',
        id: 'delete-price',
        modalTitle: 'Confirm you want to remove this price',
      }),
      {},
    );
    expect(PriceForm).toHaveBeenNthCalledWith(
      2,
      {
        action: 'delete',
        defaultValues: {
          date: '2023-01-01',
          fk_commodity: undefined,
          fk_currency: undefined,
          guid: '1',
          value: 100,
        },
      },
      {},
    );

    expect(container).toMatchSnapshot();
  });
});
