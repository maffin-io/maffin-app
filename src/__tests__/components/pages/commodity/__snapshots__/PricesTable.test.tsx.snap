// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PricesTable renders Actions column as expected 1`] = `
<div>
  <div
    data-testid="FormButton"
  >
    <div
      data-testid="PriceForm"
    />
  </div>
  <div
    data-testid="FormButton"
  >
    <div
      data-testid="PriceForm"
    />
  </div>
</div>
`;
